src/main/java/de/ait/secondlife/services/OfferServiceImpl.java

```diff
@@ -0,0 +1,157 @@
+            newOffer.setAuctionDurationDays(newOffer.getAuctionDurationDays() <= 0 ? 3 : newOffer.getAuctionDurationDays());
+            if (dto.getIsFree() == null) {
+                newOffer.setIsFree(dto.getStartPrice() == null || dto.getStartPrice().compareTo(BigDecimal.ZERO) == 0);
+            }
+            if (Boolean.TRUE.equals(newOffer.getIsFree())) {
+                checkOfferIfIsFree(dto.getStartPrice(), dto.getStep(), dto.getWinBid());
+            } else {
+                if (dto.getStartPrice() == null || dto.getStartPrice().compareTo(BigDecimal.ZERO) == 0)
+                    throw new WrongAuctionParameterException("start prise");
+                if (dto.getStep() == null || dto.getStep().compareTo(BigDecimal.ZERO) == 0)
+                    throw new WrongAuctionParameterException("step");
+                if (dto.getWinBid() != null && dto.getWinBid().compareTo(BigDecimal.ZERO) == 0) {
+                    throw new WrongAuctionParameterException("winBid");
+                }
+            }
+            newOffer = offerRepository.save(newOffer);
+            return mappingService.toRequestDto(newOffer);
+        } catch (ConstraintViolationException | DataIntegrityViolationException e) {
+            throw new CreateOfferConstraintViolationException("Constraint violation: " + e.getMessage());
+        }
+    }
+
+    @Transactional
+    @Override
+    public void updateOffer(OfferUpdateDto dto) {
+        Offer offer = offerRepository.findById(dto.getId())
+                .orElseThrow(() -> new OfferNotFoundException(dto.getId()));
+
+        offer.setTitle(dto.getTitle() == null ? offer.getTitle() : dto.getTitle());
+        offer.setDescription(dto.getDescription() == null ? offer.getDescription() : dto.getDescription());
+        offer.setAuctionDurationDays(dto.getAuctionDurationDays() == null || dto.getAuctionDurationDays() <= 0 ?
+                offer.getAuctionDurationDays() : dto.getAuctionDurationDays());
+        offer.setIsFree(dto.getIsFree() == null ? offer.getIsFree() : dto.getIsFree());
+        if (offer.getIsFree()) {
+            checkOfferIfIsFree(dto.getStartPrice(), dto.getStep(), dto.getWinBid());
+            offer.setStartPrice(null);
+            offer.setStep(null);
+            offer.setWinBid(null);
+        } else {
+            offer.setStartPrice(dto.getStartPrice() == null || dto.getStartPrice().compareTo(BigDecimal.ZERO) == 0 ?
+                    offer.getStartPrice() : dto.getStartPrice());
+            offer.setStep(dto.getStep() == null || dto.getStep().compareTo(BigDecimal.ZERO) == 0 ?
+                    offer.getStep() : dto.getStep());
+            offer.setWinBid(dto.getWinBid() == null || dto.getWinBid().compareTo(BigDecimal.ZERO) == 0 ?
+                    offer.getWinBid() : dto.getWinBid());
+        }
+    }
+
+    @Transactional
+    @Override
+    public void removeOffer(UUID id) {
```

Comment:
the code in the removeOffer and recoverOffer methods is the same. we can leave only one, rename it and send activity value through method parameters